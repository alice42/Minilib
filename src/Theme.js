export const theme = {
  backgroundColor: '#ECEFF1',
  onBackground: '#000000',
  surfaceColor: '#FFFFFF',
  surfaceColorHover: '#F5F5F5',
  surfaceColorActive: '#E0E0E0',
  onSurface: '#000000',
  primaryColor: '#0277BD',
  primaryColorHover: '#0288D1',
  primaryColorActive: '#01579B',
  onPrimary: '#FFFFFF',
  secondaryColor: '#00897B',
  secondaryColorHover: '#00695C',
  secondaryColorActive: '#004D40',
  onSecondary: '#FFFFFF',
  tertiaryColor: '#EF6C00',
  tertiaryColorHover: '#FB8C00',
  tertiaryColorActive: '#E65100',
  onTertiary: '#FFFFFF',
  fontPrimary: {
    regular: 'Lato,sans-serif',
    bold: 'Lato,sans-serif',
    light: 'Lato,sans-serif'
  },
  fontSecondary: {
    regular: 'arial,sans-serif',
    bold: 'arial,sans-serif',
    light: 'arial,sans-serif'
  },
  fontTertiary: {
    regular: 'Georgia,serif',
    bold: 'Georgia,serif',
    light: 'Georgia,serif'
  },
  font48: '4rem',
  font43: '3.6rem',
  font38: '3.2rem',
  font34: '2.9rem',
  font30: '2.5rem',
  font29: '2.4rem',
  font28: '2.3rem',
  font26: '2.1rem',
  font24: '2rem',
  font23: '1.9rem',
  font22: '1.85rem',
  font20: '1.65rem',
  font19: '1.6rem',
  font18: '1.5rem',
  font17: '1.4rem',
  font16: '1.35rem',
  font15: '1.25rem',
  font14: '1.2rem',
  font13: '1.1rem',
  font12: '1rem',
  font11: '0.95rem',
  font10: '0.9rem',
  font9: '0.75rem',
  font8: '0.7rem',
  font6: '0.5rem',
  overlayOpacity: 0.7,
  borderRadius: '5px',
  border: 'solid 5px #000000'
}
