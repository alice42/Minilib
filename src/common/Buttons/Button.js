import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { ButtonStyled } from './ButtonStyled'
import Icon from '../Icons/Icon'

export default class Button extends Component {
  static propTypes = {
    /** Label of the button */
    label: PropTypes.string,
    /** Icon of the button */
    icon: PropTypes.string,
    /** Color of the icon */
    iconColor: PropTypes.string,
    /** Size of the icon */
    iconSize: PropTypes.string,
    /** Apply a radius on corners */
    rounded: PropTypes.bool,
    /** Fill background with selected color */
    fillBackground: PropTypes.bool,
    /** Your own css propreties */
    customStyle: PropTypes.object,
    /** Action triggered when button clicked */
    onClick: PropTypes.func,
    /** Action triggered when button hoverded: default log 'Hovered !' in console */
    onMouseOver: PropTypes.func
  }

  static defaultProps = {
    label: null,
    icon: null,
    iconColor: null,
    iconSize: '2rem',
    rounded: false,
    fillBackground: false,
    customStyle: null
  }

  handleIcon = () => {
    if (this.props.icon) {
      return <Icon name={this.props.icon} iconColor={this.props.iconColor} iconSize={this.props.iconSize} />
    }
  }

  render () {
    return <ButtonStyled {...this.props} aria-label={this.props.label || this.props.icon} >{this.props.label}
      {this.handleIcon()}
    </ButtonStyled>
  }
}
