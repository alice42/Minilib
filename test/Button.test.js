import React from 'react'
import Button from '../src/common/Buttons/Button'
import 'jest-styled-components'
import { renderCreateWithTheme } from './TestUtils'
import { theme } from '../src/Theme.js'
import { shallow } from 'enzyme'
// import './classTest.css'

describe('Test Button component rendering', () => {
  it('Squared active with fill', () => {
    const component = renderCreateWithTheme(
      <Button fillBackground label='Button Label' />
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    expect(tree).toHaveStyleRule('background-color', theme.primaryColorActive)
    expect(tree).toHaveStyleRule('border-radius', 'unset')
  })

  it('Rounded active with fill', () => {
    const component = renderCreateWithTheme(
      <Button rounded fillBackground label='Button Label' />
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    expect(tree).toHaveStyleRule('background-color', theme.primaryColorActive)
    expect(tree).toHaveStyleRule('border-radius', '3.75rem')
  })

  it('Squared active with no fill', () => {
    const component = renderCreateWithTheme(<Button label='Button Label' />)
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    expect(tree).toHaveStyleRule('background-color', 'transparent')
    expect(tree).toHaveStyleRule(
      'border',
      `0.125rem solid ${theme.primaryColorActive}`
    )
    expect(tree).toHaveStyleRule('border-radius', 'unset')
  })

  it('Rounded active with no fill', () => {
    const component = renderCreateWithTheme(
      <Button rounded label='Button Label' />
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    expect(tree).toHaveStyleRule('background-color', 'transparent')
    expect(tree).toHaveStyleRule(
      'border',
      `0.125rem solid ${theme.primaryColorActive}`
    )
    expect(tree).toHaveStyleRule('border-radius', '3.75rem')
  })
  // it('With Icon', () => {
  //   const component = renderCreateWithTheme(
  //     <Button
  //       label={'Long button text'}
  //       icon={'travelType'}
  //       rounded
  //     />
  //   )
  //   let tree = component.toJSON()
  //   expect(tree).toMatchSnapshot()
  //   expect(tree).toHaveStyleRule('flex-direction', 'row')
  //   expect(tree).toHaveStyleRule('justify-content', 'space-evenly')
  //   expect(tree).toHaveStyleRule('height', 'fit-content')
  //   expect(tree).toHaveStyleRule('width', 'fit-content')
  //   expect(tree).toHaveStyleRule('text-indent', '0.625rem')
  // })
  // it('With Icon on Left', () => {
  //   const component = renderCreateWithTheme(
  //     <Button
  //       label={'Long button text '}
  //       icon={'travelType'}
  //       iconLeft
  //     />
  //   )
  //   let tree = component.toJSON()
  //   expect(tree).toMatchSnapshot()
  //   expect(tree).toHaveStyleRule('flex-direction', 'row-reverse')
  //   expect(tree).toHaveStyleRule('text-indent', 'unset')
  // })

  it('Custom Style', () => {
    const component = renderCreateWithTheme(
      <Button
        customStyle={{
          backgroundColor: 'gray',
          color: 'lightgray',
          borderColor: 'darkgray',
          textTransform: 'lowercase',
          borderBottomRightRadius: '3.75rem',
          borderTopRightRadius: '3.75rem',
          fontFamily: props => props.theme.fontSecondary.bold,
          ':hover': {
            backgroundColor: 'lightgray',
            color: 'white'
          }
        }}
      />
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    expect(tree).toHaveStyleRule('background-color', 'gray')
    expect(tree).toHaveStyleRule('border-bottom-right-radius', '3.75rem')
    expect(tree).toHaveStyleRule('border-bottom-left-radius', undefined)
  })

  /* This test fail due to css import but render properly in storybook */

  // it('Custom Class', () => {
  //   const component = renderCreateWithTheme(
  //     <Button
  //       className='classTest'
  //       label='Button Label'
  //     />
  //   )
  //   let tree = component.toJSON()
  //   expect(tree).toMatchSnapshot()
  //   expect(tree.props.className).toContain('classTest')
  //   expect(tree).toHaveStyleRule('background-color', 'aquamarine')
  // })
})

describe('Test Button component actions', () => {
  it('Test click event', () => {
    const mockOnClick = jest.fn()
    const button = shallow(<Button onClick={mockOnClick} />)
    button.simulate('click')
    expect(mockOnClick).toHaveBeenCalled()
  })
  it('Test hover event', () => {
    const mockOnMouseOver = jest.fn()
    const button = shallow(<Button onMouseOver={mockOnMouseOver} />)
    button.simulate('mouseOver')
    expect(mockOnMouseOver).toHaveBeenCalled()
  })
})
