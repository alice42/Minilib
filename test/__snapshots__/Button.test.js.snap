// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Button component rendering Custom Style 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-width: -webkit-fit-content;
  min-width: -moz-fit-content;
  min-width: fit-content;
  min-height: -webkit-fit-content;
  min-height: -moz-fit-content;
  min-height: fit-content;
  height: 2.5rem;
  width: 10rem;
  -webkit-letter-spacing: 0.0625rem;
  -moz-letter-spacing: 0.0625rem;
  -ms-letter-spacing: 0.0625rem;
  letter-spacing: 0.0625rem;
  text-align: center;
  text-transform: uppercase;
  text-indent: unset;
  padding: 0;
  padding-right: unset;
  font-family: Lato,sans-serif;
  font-size: 1rem;
  background-color: transparent;
  color: #01579B;
  border: 0.125rem solid #01579B;
  border-radius: unset;
  background-color: gray;
  color: lightgray;
  border-color: darkgray;
  text-transform: lowercase;
  border-bottom-right-radius: 3.75rem;
  border-top-right-radius: 3.75rem;
}

.c0:hover {
  background-color: #0288D1;
  color: #FFFFFF;
}

.c0:focus {
  outline: none;
}

.c0 font-family:function fontFamily(props) {
  return: props.theme.fontSecondary.bold;
}

.c0:hover {
  background-color: lightgray;
  color: white;
}

<button
  aria-label={null}
  className="c0"
  label={null}
/>
`;

exports[`Test Button component rendering Rounded active with fill 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-width: -webkit-fit-content;
  min-width: -moz-fit-content;
  min-width: fit-content;
  min-height: -webkit-fit-content;
  min-height: -moz-fit-content;
  min-height: fit-content;
  height: 2.5rem;
  width: 10rem;
  -webkit-letter-spacing: 0.0625rem;
  -moz-letter-spacing: 0.0625rem;
  -ms-letter-spacing: 0.0625rem;
  letter-spacing: 0.0625rem;
  text-align: center;
  text-transform: uppercase;
  text-indent: unset;
  padding: 0;
  padding-right: unset;
  font-family: Lato,sans-serif;
  font-size: 1rem;
  background-color: #01579B;
  color: #FFFFFF;
  border: none;
  border-radius: 3.75rem;
}

.c0:hover {
  background-color: #0288D1;
  color: #FFFFFF;
}

.c0:focus {
  outline: none;
}

<button
  aria-label="Button Label"
  className="c0"
  label="Button Label"
>
  Button Label
</button>
`;

exports[`Test Button component rendering Rounded active with no fill 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-width: -webkit-fit-content;
  min-width: -moz-fit-content;
  min-width: fit-content;
  min-height: -webkit-fit-content;
  min-height: -moz-fit-content;
  min-height: fit-content;
  height: 2.5rem;
  width: 10rem;
  -webkit-letter-spacing: 0.0625rem;
  -moz-letter-spacing: 0.0625rem;
  -ms-letter-spacing: 0.0625rem;
  letter-spacing: 0.0625rem;
  text-align: center;
  text-transform: uppercase;
  text-indent: unset;
  padding: 0;
  padding-right: unset;
  font-family: Lato,sans-serif;
  font-size: 1rem;
  background-color: transparent;
  color: #01579B;
  border: 0.125rem solid #01579B;
  border-radius: 3.75rem;
}

.c0:hover {
  background-color: #0288D1;
  color: #FFFFFF;
}

.c0:focus {
  outline: none;
}

<button
  aria-label="Button Label"
  className="c0"
  label="Button Label"
>
  Button Label
</button>
`;

exports[`Test Button component rendering Squared active with fill 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-width: -webkit-fit-content;
  min-width: -moz-fit-content;
  min-width: fit-content;
  min-height: -webkit-fit-content;
  min-height: -moz-fit-content;
  min-height: fit-content;
  height: 2.5rem;
  width: 10rem;
  -webkit-letter-spacing: 0.0625rem;
  -moz-letter-spacing: 0.0625rem;
  -ms-letter-spacing: 0.0625rem;
  letter-spacing: 0.0625rem;
  text-align: center;
  text-transform: uppercase;
  text-indent: unset;
  padding: 0;
  padding-right: unset;
  font-family: Lato,sans-serif;
  font-size: 1rem;
  background-color: #01579B;
  color: #FFFFFF;
  border: none;
  border-radius: unset;
}

.c0:hover {
  background-color: #0288D1;
  color: #FFFFFF;
}

.c0:focus {
  outline: none;
}

<button
  aria-label="Button Label"
  className="c0"
  label="Button Label"
>
  Button Label
</button>
`;

exports[`Test Button component rendering Squared active with no fill 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-width: -webkit-fit-content;
  min-width: -moz-fit-content;
  min-width: fit-content;
  min-height: -webkit-fit-content;
  min-height: -moz-fit-content;
  min-height: fit-content;
  height: 2.5rem;
  width: 10rem;
  -webkit-letter-spacing: 0.0625rem;
  -moz-letter-spacing: 0.0625rem;
  -ms-letter-spacing: 0.0625rem;
  letter-spacing: 0.0625rem;
  text-align: center;
  text-transform: uppercase;
  text-indent: unset;
  padding: 0;
  padding-right: unset;
  font-family: Lato,sans-serif;
  font-size: 1rem;
  background-color: transparent;
  color: #01579B;
  border: 0.125rem solid #01579B;
  border-radius: unset;
}

.c0:hover {
  background-color: #0288D1;
  color: #FFFFFF;
}

.c0:focus {
  outline: none;
}

<button
  aria-label="Button Label"
  className="c0"
  label="Button Label"
>
  Button Label
</button>
`;
